Mathlib/FieldTheory/Perfect.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2023 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import Mathlib.FieldTheory.Separable
+
+/-!
+
+# Perfect fields and rings
+
+In this file we define perfect fields, together with a generalisation to (commutative) rings in
+prime characteristic.
+
+## Main definitions / statements:
+ * `PerfectRing`: a ring of characteristic `p` (prime) is said to be perfect in the sense of Serre,
+   if its Frobenius map is bijective.
+ * `PerfectField`: a field `K` is said to be perfect if every irreducible polynomial over `K` is
+   separable.
+
+-/
+
+open Function
+
+/-- A perfect ring of characteristic `p` (prime) in the sense of Serre.
+
+NB: This is not related to the concept with the same name introduced by Bass (related to projective
+covers of modules). -/
+class PerfectRing (R : Type _) (p : ℕ) [CommRing R] [Fact p.Prime] [CharP R p] : Prop where
+  bijective_frobenius : Bijective $ frobenius R p
+
+section PerfectRing
+
+variable (R : Type _) (p : ℕ) [CommRing R] [Fact p.Prime] [CharP R p]
+
+/-- For a reduced ring, surjectivity of the Frobenius map is a sufficient condition for perfection.
+-/
+lemma mkOfReduced [IsReduced R] (h : Surjective $ frobenius R p) : PerfectRing R p :=
+  ⟨frobenius_inj R p, h⟩
+
+instance ofFiniteOfIsReduced [Finite R] [IsReduced R] : PerfectRing R p :=
+  mkOfReduced _ _ $ Finite.surjective_of_injective (frobenius_inj R p)
+
+variable [PerfectRing R p]
+
+@[simps!]
+noncomputable def frobeniusEquiv : R ≃+* R :=
+  RingEquiv.ofBijective (frobenius R p) PerfectRing.bijective_frobenius
+#align frobenius_equiv frobeniusEquiv
+
+@[simp]
+theorem coe_frobeniusEquiv : ⇑(frobeniusEquiv R p) = frobenius R p := rfl
+#align coe_frobenius_equiv coe_frobeniusEquiv
+
+@[simp]
+theorem frobenius_comp_frobeniusEquiv_symm :
+    (frobenius R p).comp (frobeniusEquiv R p).symm = RingHom.id R := by
+  ext; simp [surjInv_eq]
```

Comment:
I think in mathlib we should only use `simp only` but not `simp`?
```suggestion
  ext; simp only [RingHom.coe_comp, RingHom.coe_coe, comp_apply, frobeniusEquiv_symm_apply,
    surjInv_eq, RingHom.id_apply]
```

Reply:
It's OK to use `simp` to finish a proof (we call this a "terminal simp"). So I would only change this to `simp only` if it was slow but it's actually very fast so I'll leave it.